### PR TITLE
feat: rewrite test runner using libvirt

### PR DESCRIPTION
**Pull request by d-loose**
This PR replaces the runner script added in https://github.com/canonical/ubuntu-gui-testing/pull/84 with a rewrite using `libvirt` instead of manually spawning `qemu` and `swtpm` subprocesses. `yarf` is still treated as an external dependency.

Future work beyond the scope of this PR:
- add cli options for domain template modification (e.g. remove TPM)
- explore options to define test environment requirements and dependencies here in this repo.
  - yet another script that reads definitions from a yaml and invokes the runner, e.g.
    ```yaml
    - desktop-installer:
      - resolute.entire-disk:
        iso: ubuntu-26.04-desktop-amd64.iso
    - firefox-example:
      - firefox-example-basic:
        source-domain: ugt-desktop-installer-resolute.entire-disk
    ```
   - create a robotframework library with keywords to setup the test machine.
     machine setup could then be integrated into the `.robot` files, e.g.
     ```robot
     *** Settings ***
     Suite Setup  Boot ISO ubuntu-26.04-desktop-amd64.iso
     ```

  

UDENG-9899

### CODE REVIEW

File: runner/artifacts/output.xml [hunk 1]
```diff
@@ -0,0 +1,13 @@
+<?xml version="1.0" encoding="UTF-8"?>
+<robot generator="Robot 7.4.2 (Python 3.12.12 on linux)" generated="2026-05-28T16:24:22.906088" rpa="false" schemaversion="5">
+<suite id="s1" name="desktop-security-center" source="/tmp/tmpwspt3lw1">
+<suite id="s1-s1" name="Desktop-Security-Center-Check-Key-Failure" source="/tmp/tmpwspt3lw1/desktop-security-center-check-key-failure">
+<suite id="s1-s1-s1" name="Desktop-Security-Center-Check-Key-Failure" source="/tmp/tmpwspt3lw1/desktop-security-center-check-key-failure/desktop-security-center-check-key-failure.robot">
+<test id="s1-s1-s1-t1" name="Log In" line="14">
+<kw name="Log In" owner="common">
+<kw name="Match Text" owner="VideoInput">
+<msg time="2026-05-28T16:24:45.542307" level="INFO">Looking for 'Not listed'
```
Comment (adombeck) [inline runner/artifacts/output.xml]: was this file committed on purpose?

File: runner/ubuntu_gui_testing_runner/base.py [hunk 2]
```diff
@@ -0,0 +1,272 @@
...
+            "--outdir",
+            "./artifacts",
+            "--",
+            "--variable",
+            f"CID:{vsock_cid}",
+            "--suite",
+            test,
+        ]
+        env = os.environ.copy()
+        env["VNC_PORT"] = str(vnc_port - 5900)
```
Comment (adombeck) [inline runner/ubuntu_gui_testing_runner/base.py:248]: should we handle the case that vnc_port < 5900?

File: runner/ubuntu_gui_testing_runner/iso_runner.py [hunk 3]
```diff
@@ -0,0 +1,168 @@
...
+                guest_cid=vsock_cid,
+                test_username=self.test_username,
+                test_password=self.test_password,
+                artifacts_dir=self.artifacts_path,
+            )
+
+    async def _handle_reboot(self) -> None:
+        """Poll until the transient domain shuts down, then start the persistent one."""
+        while self.domain.isActive():
+            await asyncio.sleep(1)
```
Comment (adombeck) [inline runner/ubuntu_gui_testing_runner/iso_runner.py:168]: might be overengineering, but it is possible to wait for domain lifecycle events without polling by using [virConnectDomainEventRegisterAny](https://libvirt.org/html/libvirt-libvirt-domain.html#virConnectDomainEventRegisterAny) with [`VIR_DOMAIN_EVENT_ID_LIFECYCLE`](https://libvirt.org/html/libvirt-libvirt-domain.html#virDomainEventID)

Comment (adombeck) [reply runner/ubuntu_gui_testing_runner/iso_runner.py:168]: (I'd probably stick with the sleep for now)

File: runner/ubuntu_gui_testing_runner/iso_runner.py [hunk 4]
```diff
@@ -0,0 +1,168 @@
...
+        disk_volume_xml = self._render_template(
+            self.volume_template_path, {"name": self.disk_volume_name}
+        )
+        self._disk_volume = self.pool.createXML(disk_volume_xml, 0)
+
+    def _create_domain(self) -> None:
+        # Phase 1: Create a transient domain that boots from the ISO with
+        # on_reboot=destroy.  When the installer finishes and triggers a
+        # reboot the domain will shut down instead of rebooting, giving us
+        # a signal that installation is complete.
```
Comment (adombeck) [inline runner/ubuntu_gui_testing_runner/iso_runner.py:70]: What if a test case (beside the installer) requires the VM to reboot? Probably a rare edge case which we can deal with once we encounter it.

**Review by adombeck — COMMENTED** [review]
Looks good! Just some minor comments.

File: runner/artifacts/output.xml (hunk 1 above)
Comment (adombeck) [reply runner/artifacts/output.xml]: ah, already removed :)

File: runner/ubuntu_gui_testing_runner/iso_runner.py (hunk 4 above)
Comment (d-loose) [reply runner/ubuntu_gui_testing_runner/iso_runner.py:70]: For tests using the `LibvirtImageRunner` that's not a problem, reboots work perfectly fine. The `LibvirtIsoRunner` currently assumes that we need to remove the CD during the first reboot. If there are test cases that boot from a live ISO and need to perform a reboot without removing the CD, we could add a cli flag for that

Comment (adombeck) [reply runner/ubuntu_gui_testing_runner/iso_runner.py:70]: I was expecting the `on_reboot=destroy` flag to cause the VM to not reboot.

Comment (d-loose) [reply runner/ubuntu_gui_testing_runner/iso_runner.py:70]: The image runner has its own domain template where this is set to `restart`:
https://github.com/canonical/ubuntu-gui-testing/blob/3d1cba2a56bd760ec02d6c600d27232c572d341f/runner/templates/image_domain_template.xml#L28

Comment (adombeck) [reply runner/ubuntu_gui_testing_runner/iso_runner.py:70]: yes, but for the ISO runner it's set to `destroy`, so there any reboot will cause the VM to be destroyed (i.e. stopped), right? so the issue with rebooting is not only the removal of the CD. in any case, it seems like enough of an edge case that I don't want to urge you handle it now. just thought it might be worth noting.

**Review by adombeck — APPROVED** [review]
